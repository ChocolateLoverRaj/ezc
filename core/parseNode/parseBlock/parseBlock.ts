import CoreTokenDatas from '../../CoreTokenDatas'
import CoreTokenType from '../../CoreTokenType'
import EnumItemWithData from '../../EnumItemWithData'
import IdentifierType from '../../IdentifierType'
import skipSplittedIterator from '../../../util/splitAsyncIterator/skip'
import splitAsyncIterator from '../../../util/splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'
import TryParseNode from '../TryParseNode'

// TODO: Validate if block ends with a instruction that switches to a different block
const parseBlock = (
  instructionParsers: ReadonlyArray<TryParseNode<EnumItemWithData>>
): TryParseNode<CoreNodesWithData[CoreNodeType.BLOCK]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
  let length = 0
  const skip = (count: number): void => {
    length += count
    skipSplittedIterator(splittedIterator, count)
  }
  const type = {
    enum: CoreNodeType,
    id: CoreNodeType.BLOCK
  }

  const name = await (async () => {
    const { done, value } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.IDENTIFIER)) return
    const { name, type } = value.data as CoreTokenDatas[CoreTokenType.IDENTIFIER]
    if (type !== IdentifierType.BLOCK) return
    skip(1)
    return name
  })()
  if (name === undefined) {
    return {
      success: false,
      result: {
        type,
        index: 0,
        message: 'Expected name identifier',
        subAttempts: undefined
      }
    }
  }

  const instructions: EnumItemWithData[] = []
  while (true) {
    const { success, result } = await tryNodeParsers(instructionParsers)(
      splittedIterator.asyncIterable)
    if (!success) break
    instructions.push(result.node)
    skip(result.length)
  }

  return {
    success: true,
    result: {
      node: {
        type,
        data: {
          name,
          instructions
        }
      },
      length
    }
  }
}

export default parseBlock
