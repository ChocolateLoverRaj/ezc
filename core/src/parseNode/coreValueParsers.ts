import EnumItemWithData from '../EnumItemWithData'
import parseIdentifier from './parseIdentifier'
import parseNumber from './parseNumber'
import parseString from './parseString'
import TryParseNode from './TryParseNode'

const coreValueParsers: ReadonlyArray<TryParseNode<EnumItemWithData>> = [
  parseString,
  parseNumber,
  parseIdentifier
]

export default coreValueParsers
