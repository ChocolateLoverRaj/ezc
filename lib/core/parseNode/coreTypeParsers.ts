import parseArrayType from './parseArrayType'
import parseIntegerType from './parseIntegerType'
import parsePointerType from './parsePointerType'
import parseVoidType from './parseVoidType'
import TypeParser from './TypeParser'

const coreTypeParsers = ((): readonly TypeParser[] => {
  const coreTypeParsers: TypeParser[] = [
    parseIntegerType,
    parsePointerType,
    parseVoidType
  ]
  coreTypeParsers.push(parseArrayType(coreTypeParsers))
  return coreTypeParsers
})()

export default coreTypeParsers
