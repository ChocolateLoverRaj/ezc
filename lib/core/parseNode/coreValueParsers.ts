import EnumItemWithData from '../EnumItemWithData'
import parseNumber from './parseNumber'
import parseString from './parseString'
import TryParseNode from './TryParseNode'

const coreValueParsers: ReadonlyArray<TryParseNode<EnumItemWithData>> = [
  parseString,
  parseNumber
]

export default coreValueParsers
