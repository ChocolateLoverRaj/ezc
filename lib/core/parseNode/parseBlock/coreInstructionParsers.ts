import EnumItemWithData from '../../EnumItemWithData'
import coreInput from '../parseReturnInstruction/coreInput'
import parseReturnInstruction from '../parseReturnInstruction/parseReturnInstruction'
import TryParseNode from '../TryParseNode'

const coreInstructionParsers: ReadonlyArray<TryParseNode<EnumItemWithData>> = [
  parseReturnInstruction(coreInput)
]

export default coreInstructionParsers
