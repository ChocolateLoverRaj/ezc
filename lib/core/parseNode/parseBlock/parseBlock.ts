import CoreKeyWord from '../../CoreKeyWord'
import CoreTokensWithData from '../../CoreTokensWithData'
import CoreTokenType from '../../CoreTokenType'
import skip from '../../splitAsyncIterator/skip'
import splitAsyncIterator from '../../splitAsyncIterator/splitAsyncIterator'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import tryNodeParsers from '../tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'

const parseBlock = (
  { typeParsers, valueParsers }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.RETURN_INSTRUCTION]> => async stream => {
  const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())

  {
    const { value, done } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const { data } = value as CoreTokensWithData[CoreTokenType.KEY_WORD]
    if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.RETURN)) return
    skip(splittedIterator, 1)
  }

  const parsedType = await tryNodeParsers(typeParsers)(splittedIterator.asyncIterable)
  if (parsedType === undefined) return
  skip(splittedIterator, parsedType.length)

  const parsedValue = await tryNodeParsers(valueParsers)(splittedIterator.asyncIterable)
  if (parsedValue === undefined) return
  skip(splittedIterator, parsedValue.length)

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.RETURN_INSTRUCTION
      },
      data: {
        type: parsedType.node,
        value: parsedValue.node
      }
    },
    length: 1 + parsedType.length + parsedValue.length
  }
}

export default parseBlock
