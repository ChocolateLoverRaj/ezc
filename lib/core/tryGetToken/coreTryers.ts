import CoreTokenWithData from '../CoreTokenWithData'
import coreParseKeywordOptions from './coreParseKeywordOptions'
import parseIdentifier from './parseIdentifier'
import parseIntegerType from './parseIntegerType'
import parseKeyword from './parseKeyword'
import parseNumberLiteral from './parseNumberLiteral'
import parseOpenClose from './parseOpenClose'
import parseStringLiteral from './parseStringLiteral'
import TryGetToken from './TryGetToken'

const coreTryers: ReadonlyArray<TryGetToken<CoreTokenWithData>> = [
  parseIdentifier,
  parseIntegerType,
  parseKeyword(coreParseKeywordOptions),
  parseNumberLiteral,
  parseOpenClose,
  parseStringLiteral
] as const

export default coreTryers
