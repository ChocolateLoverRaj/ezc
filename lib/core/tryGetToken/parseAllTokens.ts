import EnumItemWithData from '../EnumItemWithData'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import parseToken from './parseToken'
import TryGetToken from './TryGetToken'

const parseAllTokens = (tryers: ReadonlyArray<TryGetToken<EnumItemWithData>>) => async function * (
  asyncIterable: AsyncIterable<string>
): AsyncGenerator<EnumItemWithData | undefined> {
  const splittedIterator = splitAsyncIterator(asyncIterable[Symbol.asyncIterator]())
  let chunkBefore = ''
  while (true) {
    const token = await parseToken(tryers)({
      async * [Symbol.asyncIterator] () {
        yield chunkBefore
        for await (const chunk of splittedIterator.asyncIterable) {
          yield chunk
        }
      }
    })
    if (token === undefined) {
      yield undefined
      return
    }
    yield token.token
    let chunksProcessed = 0
    let charactersProcessed = 0
    if (token.length < chunkBefore.length) {
      chunkBefore = chunkBefore.slice(token.length - chunkBefore.length)
    } else {
      for await (const chunk of splittedIterator.asyncIterable) {
        chunksProcessed++
        charactersProcessed += chunk.length
        if (charactersProcessed >= token.length) {
          chunkBefore = charactersProcessed === token.length
            ? ''
            : chunk.slice(token.length - charactersProcessed)
          break
        }
      }
      if (chunksProcessed === 0) return
      skip(splittedIterator, chunksProcessed)
    }
  }
}

export default parseAllTokens
