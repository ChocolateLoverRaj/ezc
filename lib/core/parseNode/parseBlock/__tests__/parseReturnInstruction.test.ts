import arrayToAsyncIterable from '../../../arrayToAsyncIterable'
import coreTryers from '../../../tryGetToken/coreTryers'
import parseAllTokens from '../../../tryGetToken/parseAllTokens'
import coreTypeParsers from '../../coreTypeParsers'
import coreValueParsers from '../../coreValueParsers'
import parseBlock from '../parseBlock'

test('ret i1 0', async () => {
  await expect(parseBlock({
    typeParsers: coreTypeParsers,
    valueParsers: coreValueParsers
  })(parseAllTokens(coreTryers)(arrayToAsyncIterable([
    'ret i1 0'
  ])) as any)).resolves.toMatchSnapshot()
})
