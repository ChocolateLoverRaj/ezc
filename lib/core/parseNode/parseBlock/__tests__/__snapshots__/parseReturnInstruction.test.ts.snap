// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ret i1 0 1`] = `
Object {
  "length": 3,
  "node": Object {
    "data": Object {
      "type": Object {
        "data": 1,
        "type": Object {
          "enum": Object {
            "0": "FILE",
            "1": "DECLARE",
            "10": "IDENTIFIER",
            "11": "NUMBER",
            "12": "VOID_TYPE",
            "13": "BLOCK",
            "14": "RETURN_INSTRUCTION",
            "2": "FUNCTION",
            "3": "INTEGER_TYPE",
            "4": "GLOBAL_VARIABLE",
            "5": "STRING",
            "6": "FLOAT_TYPE",
            "7": "POINTER_TYPE",
            "8": "ARRAY_TYPE",
            "9": "STRUCT_TYPE",
            "ARRAY_TYPE": 8,
            "BLOCK": 13,
            "DECLARE": 1,
            "FILE": 0,
            "FLOAT_TYPE": 6,
            "FUNCTION": 2,
            "GLOBAL_VARIABLE": 4,
            "IDENTIFIER": 10,
            "INTEGER_TYPE": 3,
            "NUMBER": 11,
            "POINTER_TYPE": 7,
            "RETURN_INSTRUCTION": 14,
            "STRING": 5,
            "STRUCT_TYPE": 9,
            "VOID_TYPE": 12,
          },
          "id": 3,
        },
      },
      "value": Object {
        "data": 0,
        "type": Object {
          "enum": Object {
            "0": "FILE",
            "1": "DECLARE",
            "10": "IDENTIFIER",
            "11": "NUMBER",
            "12": "VOID_TYPE",
            "13": "BLOCK",
            "14": "RETURN_INSTRUCTION",
            "2": "FUNCTION",
            "3": "INTEGER_TYPE",
            "4": "GLOBAL_VARIABLE",
            "5": "STRING",
            "6": "FLOAT_TYPE",
            "7": "POINTER_TYPE",
            "8": "ARRAY_TYPE",
            "9": "STRUCT_TYPE",
            "ARRAY_TYPE": 8,
            "BLOCK": 13,
            "DECLARE": 1,
            "FILE": 0,
            "FLOAT_TYPE": 6,
            "FUNCTION": 2,
            "GLOBAL_VARIABLE": 4,
            "IDENTIFIER": 10,
            "INTEGER_TYPE": 3,
            "NUMBER": 11,
            "POINTER_TYPE": 7,
            "RETURN_INSTRUCTION": 14,
            "STRING": 5,
            "STRUCT_TYPE": 9,
            "VOID_TYPE": 12,
          },
          "id": 11,
        },
      },
    },
    "type": Object {
      "enum": Object {
        "0": "FILE",
        "1": "DECLARE",
        "10": "IDENTIFIER",
        "11": "NUMBER",
        "12": "VOID_TYPE",
        "13": "BLOCK",
        "14": "RETURN_INSTRUCTION",
        "2": "FUNCTION",
        "3": "INTEGER_TYPE",
        "4": "GLOBAL_VARIABLE",
        "5": "STRING",
        "6": "FLOAT_TYPE",
        "7": "POINTER_TYPE",
        "8": "ARRAY_TYPE",
        "9": "STRUCT_TYPE",
        "ARRAY_TYPE": 8,
        "BLOCK": 13,
        "DECLARE": 1,
        "FILE": 0,
        "FLOAT_TYPE": 6,
        "FUNCTION": 2,
        "GLOBAL_VARIABLE": 4,
        "IDENTIFIER": 10,
        "INTEGER_TYPE": 3,
        "NUMBER": 11,
        "POINTER_TYPE": 7,
        "RETURN_INSTRUCTION": 14,
        "STRING": 5,
        "STRUCT_TYPE": 9,
        "VOID_TYPE": 12,
      },
      "id": 14,
    },
  },
}
`;
